### ISSUE TITLE
[Question] LIM: Wayland?

### ISSUE THREAD

**Issue by chippmann** [Zren/material-decoration#6]
Is the lim branch also compatible with wayland?

**Comment by Zren** [comment]
No. There's a few `#ifdef HAVE_X11` that I've commented out to simplify porting appmenu from `plasma-workspace`.

* https://github.com/Zren/material-decoration/blob/lim/src/AppMenuModel.cc#L28
* https://github.com/KDE/plasma-workspace/tree/master/appmenu
* https://github.com/psifidotos/applet-window-appmenu/blob/master/plugin/appmenumodel.cpp

**Comment by chippmann** [comment]
Ah bummer. Anyway thanks a lot for your work! I keep watching.
How big of a task do you think will the wayland support be (i might have a look into it once i have a bit more time)

**Comment by Zren** [comment]
If the widget works in walyand, then it should be possible.

The hard part is testing it, as while you can restart `plasmashell` easily in wayland, you can't really easily test `kwin_wayland` as killing it will restart your session.

**Comment by baduhai** [comment]
As of earlier today, when Plasma 5.19 was released, the Global Menu works under Wayland. From my testing, LIM does not work as it currently is under Wayland, but perhaps due to the advancements it could with a bit of tweaking? The global menu is currently the only thing keeping me on X11, so it'd be awesome to be able to migrate.

**Comment by trmdi** [comment]
FWIW, applet-window-appmenu has added wayland support recently.

**Comment by Zren** [comment]
Relevant changes:

* https://github.com/psifidotos/applet-window-appmenu/compare/83e19b38cf952241c3994234348f53504febfd80...master
-----
* https://github.com/psifidotos/applet-window-appmenu/commit/83e19b38cf952241c3994234348f53504febfd80
* https://github.com/psifidotos/applet-window-appmenu/commit/1b08fc22e8d8d73a07e4c0ad6405f9db8c4560df
-----
* https://github.com/psifidotos/applet-window-appmenu/commit/2799bfd7aad973f599f66532f31b8f27e478b89e
* https://github.com/psifidotos/applet-window-appmenu/commit/63f820ee571576e2f51f1e3df7dee60c70d7fda8
* https://github.com/psifidotos/applet-window-appmenu/commit/72d5b3f8f3e8d8793796b27563ea6937be0f2f5b
* https://github.com/psifidotos/applet-window-appmenu/commit/01f66220967b21b78b6c7cac21b2ed0ccf623fb5
* https://github.com/psifidotos/applet-window-appmenu/commit/6857f034e7a98635e1c92e708b2ce6d0d5257bff

**Comment by Zren** [comment]
Here's the commit for how the Global Menu widget implemented LibTaskManager.

* https://github.com/kde/plasma-workspace/commit/d8d079e42e62e2133477470c1d6b7c4559bcee55

I found I could link with `PW::LibTaskManager` or `taskmanager`. I needed to use `#include <taskmanager/tasksmodel.h>` to import the `TasksModel` class.

**Comment by Zren** [comment]
All I really need is to connect to the wayland server, find the `PlasmaWindow` instance, then get it's `applicationMenuServiceName()` and `applicationMenuObjectPath()`.

* https://github.com/KDE/plasma-workspace/blob/master/libtaskmanager/waylandtasksmodel.cpp#L122
* https://github.com/KDE/kwayland/blob/master/src/client/plasmawindowmanagement.h#L247

`PlasmaWindow::uuid()` was added in the most recent KF5 v5.73.

psifidotos'  appmenu widget has a nice `WaylandWindowManager::windowFor(QVariant wid)` I can use. It checks if `PlasmaWindow::isValid() && PlasmaWindow::internalId() == wid.toUInt()`. Unfortunately, `internalId()` is deprecated in the latest KF5 v5.73.

* https://github.com/psifidotos/applet-window-appmenu/compare/1b54d13c23cb4964566aa0510aca9a29aa1af96f...master#diff-91738585b755500885894ad0c046669dR104

**Comment by baduhai** [comment]
Seems like the replacement for `internalId()` is `uuid()`, which [doesn't seem so hard to adapt](https://api.kde.org/frameworks/kwayland/html/classKWayland_1_1Client_1_1PlasmaWindow.html#a2fcd465c75794d1b98c07cff5582c06c).

**Comment by Zren** [comment]
For some reason, wayland menus show up with a titlebar. Even the normal AppMenu icon button.

![2021-03-15___21-49-11](https://user-images.githubusercontent.com/416367/111244075-9853f000-85d8-11eb-8050-d53850643e95.png)

Oddly, the AppIcon button menu displays properly.

* `decoration->requestShowWindowMenu` https://invent.kde.org/plasma/kdecoration/-/blob/master/src/decorationbutton.cpp#L58-62
* `DecoratedClientImpl::requestShowWindowMenu` https://invent.kde.org/plasma/kwin/-/blob/master/src/decorations/decoratedclient.cpp#L229
* `Workspace::showWindowMenu` https://invent.kde.org/plasma/kwin/-/blob/master/src/useractions.cpp#L1539
* `UserActionsMenu::show()` https://invent.kde.org/plasma/kwin/-/blob/master/src/useractions.cpp#L117
* `UserActionsMenu::init()`

```cpp
m_menu = new QMenu;
// ...
m_menu->setPalette(m_client->palette());
```

The `ApplicationMenu` has a similar call:

* `Workspace::showApplicationMenu` https://invent.kde.org/plasma/kwin/-/blob/master/src/useractions.cpp#L1544
* `ApplicationMenu::showApplicationMenu` https://invent.kde.org/plasma/kwin/-/blob/master/src/appmenu.cpp#L103
* Related code: https://invent.kde.org/plasma/kwin/-/commit/93938d60b84df81733da7b3f697332fc76fe085b2

**Comment by baduhai** [comment]
5.22 fixed a lot of issues with the global menu on wayland, perhaps there could be some progress now?

That search feature is sweet, and with LIM, it'd be the ultimate setup.

**Comment by Vistaus** [comment]
I've recently switchet to Wayland and this feature would be so, so sweet to have! 🙂

**Comment by Vistaus** [comment]
@Zren Have you had time to look at this yet?

**Comment by amorozov** [comment]
Hello! Maybe some help is needed?

**Comment by flukejones** [comment]
Actually came here just to ping like the folks above. Recently shifted from Gnome to KDE and the absolute waste of space that is a titlebar is doing my head in.

I'm unlikely to find time to help unfortunately.

**Comment by flukejones** [comment]
Also came across this - https://invent.kde.org/plasma/breeze/-/merge_requests/126, is it the same code as this?

**Comment by amorozov** [comment]
Well, as the annotation says, it's based on the material decoration code. Likely its code slightly differs from the original one.  Also it doesn't support the continuation button yet.

**Comment by guiodic** [comment]
By the end of the year we will move to Plasma 6 which will have Wayland by default. It would be nice to have the LIM ported to Plasma 6 and if possible to Wayland.

**Comment by guiodic** [comment]
Hi, now my (friendly) fork works on Wayland too
https://github.com/guiodic/material-decoration
